### ISSUE TITLE
CephFS on custom cluster

### ISSUE THREAD

**Issue by penumbra23** [rook/rook#4727]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

Monitors not starting on all nodes, OSD-prepare pods not starting at all

**Expected behavior:**

Monitors and OSD pods started and running

**How to reproduce it (minimal and precise):**

1. Create a VM on AWS and one on Google Cloud
2. Join them in one VPN (with Amazon VPC, Customer Gateway and Client VPN) until they can ping each other
3. Create kube resources (`common`, `operator`, `cluster-test`)

**NOTE**:

When deploying CephFS only on AWS nodes, it works. When deploying on GCP nodes, it works. Only when creating a mixed node cluster, then it fails to deploy OSDs, MDSs, MGRs, etc. 

On Rook Slack they were suspicious about the network connectivity, but all ports are open and they see each other locally. When deploying test pods just to check if they're connected, everything works fine.

**File(s) to submit**:

Monitor logs:
```
debug 2020-01-21 11:24:11.916 7f1aa06591c0  0 starting mon.a rank 0 at public addrs [v2:10.43.57.20:3300/0,v1:10.43.57.20:6789/0] at bind addrs [v2:10.42.2.6:3300/0,v1:10.42.2.6:6789/0] mon_data /var/lib/ceph/mon/ceph-a fsid 93e3b01b-a056-4ed0-b6f9-31e607e0b4e0
debug 2020-01-21 11:24:11.916 7f1aa06591c0  1 mon.a@-1(???) e0 preinit fsid 93e3b01b-a056-4ed0-b6f9-31e607e0b4e0
debug 2020-01-21 11:24:11.916 7f1aa06591c0  1 mon.a@-1(???) e0  initial_members a, filtering seed monmap
debug 2020-01-21 11:24:11.920 7f1aa06591c0  0 mon.a@-1(probing) e0  my rank is now 0 (was -1)
debug 2020-01-21 11:24:11.920 7f1aa06591c0  1 mon.a@0(probing) e0 win_standalone_election
debug 2020-01-21 11:24:11.920 7f1aa06591c0  1 mon.a@0(probing).elector(0) init, first boot, initializing epoch at 1
debug 2020-01-21 11:24:11.920 7f1aa06591c0 -1 mon.a@0(electing) e0 failed to get devid for : udev_device_new_from_subsystem_sysname failed on ''
debug 2020-01-21 11:24:11.920 7f1aa06591c0  0 log_channel(cluster) log [INF] : mon.a is new leader, mons a in quorum (ranks 0)
debug 2020-01-21 11:24:11.924 7f1aa06591c0  1 mon.a@0(leader).osd e0 create_pending setting backfillfull_ratio = 0.9
debug 2020-01-21 11:24:11.924 7f1aa06591c0  1 mon.a@0(leader).osd e0 create_pending setting full_ratio = 0.95
debug 2020-01-21 11:24:11.924 7f1aa06591c0  1 mon.a@0(leader).osd e0 create_pending setting nearfull_ratio = 0.85
debug 2020-01-21 11:24:11.924 7f1aa06591c0  1 mon.a@0(leader).osd e0 do_prune osdmap full prune enabled
debug 2020-01-21 11:24:11.924 7f1aa06591c0  1 mon.a@0(leader).osd e0 encode_pending skipping prime_pg_temp; mapping job did not start
debug 2020-01-21 11:24:11.924 7f1aa06591c0  1 mon.a@0(leader) e0 _apply_compatset_features enabling new quorum features: compat={},rocompat={},incompat={4=support erasure code pools,5=new-style osdmap encoding,6=support isa/lrc erasure code,7=support shec erasure code}
debug 2020-01-21 11:24:11.928 7f1a86f0d700  1 mon.a@0(leader).paxosservice(auth 0..0) refresh upgraded, format 3 -> 0
debug 2020-01-21 11:24:11.932 7f1a86f0d700  1 mon.a@0(probing) e1 win_standalone_election
debug 2020-01-21 11:24:11.932 7f1a86f0d700  1 mon.a@0(probing).elector(2) init, last seen epoch 2
debug 2020-01-21 11:24:11.932 7f1a86f0d700 -1 mon.a@0(electing) e1 failed to get devid for : udev_device_new_from_subsystem_sysname failed on ''
debug 2020-01-21 11:24:11.932 7f1a86f0d700  0 log_channel(cluster) log [INF] : mon.a is new leader, mons a in quorum (ranks 0)
debug 2020-01-21 11:24:11.932 7f1a86f0d700  0 log_channel(cluster) log [DBG] : monmap e1: 1 mons at {a=[v2:10.43.57.20:3300/0,v1:10.43.57.20:6789/0]}
debug 2020-01-21 11:24:11.936 7f1a86f0d700  1 mon.a@0(leader).osd e0 create_pending setting backfillfull_ratio = 0.9
debug 2020-01-21 11:24:11.936 7f1a86f0d700  1 mon.a@0(leader).osd e0 create_pending setting full_ratio = 0.95
debug 2020-01-21 11:24:11.936 7f1a86f0d700  1 mon.a@0(leader).osd e0 create_pending setting nearfull_ratio = 0.85
debug 2020-01-21 11:24:11.936 7f1a86f0d700  1 mon.a@0(leader).osd e0 do_prune osdmap full prune enabled
debug 2020-01-21 11:24:11.936 7f1a86f0d700  1 mon.a@0(leader).osd e0 encode_pending skipping prime_pg_temp; mapping job did not start
debug 2020-01-21 11:24:11.936 7f1a86f0d700  1 mon.a@0(leader) e1 _apply_compatset_features enabling new quorum features: compat={},rocompat={},incompat={8=support monmap features,9=luminous ondisk layout,10=mimic ondisk layout,11=nautilus ondisk layout}
debug 2020-01-21 11:24:11.940 7f1a86f0d700  0 mon.a@0(leader).mds e1 new map
debug 2020-01-21 11:24:11.940 7f1a86f0d700  0 mon.a@0(leader).mds e1 print_map
e1
enable_multiple, ever_enabled_multiple: 0,0
compat: compat={},rocompat={},incompat={1=base v0.20,2=client writeable ranges,3=default file layouts on dirs,4=dir inode in separate object,5=mds uses versioned encoding,6=dirfrag is stored in omap,8=no anchor table,9=file layout v2,10=snaprealm v2}
legacy client fscid: -1

No filesystems configured

debug 2020-01-21 11:24:11.940 7f1a86f0d700  1 mon.a@0(leader).paxosservice(auth 0..0) refresh upgraded, format 3 -> 0
debug 2020-01-21 11:24:11.940 7f1a86f0d700  0 log_channel(cluster) log [DBG] : fsmap
debug 2020-01-21 11:24:11.944 7f1a86f0d700  1 mon.a@0(leader).osd e1 e1: 0 total, 0 up, 0 in
debug 2020-01-21 11:24:11.944 7f1a86f0d700  0 mon.a@0(leader).osd e1 crush map has features 288514050185494528, adjusting msgr requires
debug 2020-01-21 11:24:11.944 7f1a86f0d700  0 mon.a@0(leader).osd e1 crush map has features 288514050185494528, adjusting msgr requires
debug 2020-01-21 11:24:11.944 7f1a86f0d700  0 mon.a@0(leader).osd e1 crush map has features 3314932999778484224, adjusting msgr requires
debug 2020-01-21 11:24:11.944 7f1a86f0d700  0 mon.a@0(leader).osd e1 crush map has features 288514050185494528, adjusting msgr requires
cluster 2020-01-21 11:24:11.941083 mon.a (mon.0) 0 :  [INF] mkfs 93e3b01b-a056-4ed0-b6f9-31e607e0b4e0
debug 2020-01-21 11:24:11.944 7f1a86f0d700  1 mon.a@0(leader).paxosservice(auth 1..1) refresh upgraded, format 0 -> 3
debug 2020-01-21 11:24:11.948 7f1a86f0d700  0 log_channel(cluster) log [DBG] : osdmap e1: 0 total, 0 up, 0 in
debug 2020-01-21 11:24:11.948 7f1a86f0d700  0 log_channel(cluster) log [DBG] : mgrmap e1: no daemons active
cluster 2020-01-21 11:24:11.926808 mon.a (mon.0) 1 : cluster [INF] mon.a is new leader, mons a in quorum (ranks 0)
cluster 2020-01-21 11:24:11.937981 mon.a (mon.0) 2 : cluster [INF] mon.a is new leader, mons a in quorum (ranks 0)
cluster 2020-01-21 11:24:11.939515 mon.a (mon.0) 3 : cluster [DBG] monmap e1: 1 mons at {a=[v2:10.43.57.20:3300/0,v1:10.43.57.20:6789/0]}
cluster 2020-01-21 11:24:11.945079 mon.a (mon.0) 4 : cluster [DBG] fsmap
cluster 2020-01-21 11:24:11.953084 mon.a (mon.0) 5 : cluster [DBG] osdmap e1: 0 total, 0 up, 0 in
cluster 2020-01-21 11:24:11.954649 mon.a (mon.0) 6 : cluster [DBG] mgrmap e1: no daemons active
```

Operator logs:
```
2020-01-21 11:22:34.840159 I | cephcmd: starting operator
2020-01-21 11:22:34.897183 I | op-discover: rook-discover daemonset started
2020-01-21 11:22:34.901688 I | operator: rook-provisioner ceph.rook.io/block started using ceph.rook.io flex vendor dir
I0121 11:22:34.901927       7 leaderelection.go:217] attempting to acquire leader lease  rook-ceph/ceph.rook.io-block...
2020-01-21 11:22:34.902405 I | operator: rook-provisioner rook.io/block started using rook.io flex vendor dir
2020-01-21 11:22:34.902424 I | operator: Watching all namespaces for cluster CRDs
2020-01-21 11:22:34.902433 I | op-cluster: start watching clusters in all namespaces
2020-01-21 11:22:34.902475 I | op-cluster: Enabling hotplug orchestration: ROOK_DISABLE_DEVICE_HOTPLUG=false
I0121 11:22:34.902487       7 leaderelection.go:217] attempting to acquire leader lease  rook-ceph/rook.io-block...
2020-01-21 11:22:34.903644 I | operator: setting up the controller-runtime manager
I0121 11:22:34.922720       7 leaderelection.go:227] successfully acquired lease rook-ceph/rook.io-block
I0121 11:22:34.924002       7 event.go:209] Event(v1.ObjectReference{Kind:"Endpoints", Namespace:"rook-ceph", Name:"rook.io-block", UID:"a8b3208a-d8bf-411f-a33b-9807018ed59b", APIVersion:"v1", ResourceVersion:"1422", FieldPath:""}): type: 'Normal' reason: 'LeaderElection' rook-ceph-operator-c8ff6447d-gjgjs_5290d04e-3c40-11ea-aa30-0a580a2a0002 became leader
I0121 11:22:34.924973       7 controller.go:769] Starting provisioner controller rook.io/block_rook-ceph-operator-c8ff6447d-gjgjs_5290d04e-3c40-11ea-aa30-0a580a2a0002!
I0121 11:22:34.925011       7 leaderelection.go:227] successfully acquired lease rook-ceph/ceph.rook.io-block
I0121 11:22:34.925429       7 controller.go:769] Starting provisioner controller ceph.rook.io/block_rook-ceph-operator-c8ff6447d-gjgjs_5290a12a-3c40-11ea-aa30-0a580a2a0002!
I0121 11:22:34.925522       7 event.go:209] Event(v1.ObjectReference{Kind:"Endpoints", Namespace:"rook-ceph", Name:"ceph.rook.io-block", UID:"ade68e12-20b5-409c-a8b3-e283ad8f9c06", APIVersion:"v1", ResourceVersion:"1424", FieldPath:""}): type: 'Normal' reason: 'LeaderElection' rook-ceph-operator-c8ff6447d-gjgjs_5290a12a-3c40-11ea-aa30-0a580a2a0002 became leader
2020-01-21 11:22:34.955654 I | operator: starting the controller-runtime manager
I0121 11:22:35.725634       7 controller.go:818] Started provisioner controller ceph.rook.io/block_rook-ceph-operator-c8ff6447d-gjgjs_5290a12a-3c40-11ea-aa30-0a580a2a0002!
I0121 11:22:36.125340       7 controller.go:818] Started provisioner controller rook.io/block_rook-ceph-operator-c8ff6447d-gjgjs_5290d04e-3c40-11ea-aa30-0a580a2a0002!
2020-01-21 11:23:38.271023 I | op-cluster: starting cluster in namespace rook-ceph
2020-01-21 11:23:38.400267 I | operator: successfully started Ceph CSI driver(s)
2020-01-21 11:23:44.400667 I | op-cluster: detecting the ceph image version for image ceph/ceph:v14.2.4-20190917...
2020-01-21 11:23:51.051089 I | op-cluster: Detected ceph image version: 14.2.4 nautilus
2020-01-21 11:23:51.056924 E | cephconfig: clusterInfo: <nil>
2020-01-21 11:23:51.056949 I | op-cluster: CephCluster rook-ceph status: Creating.
2020-01-21 11:23:51.171780 I | op-mon: start running mons
2020-01-21 11:23:51.196106 I | exec: Running command: ceph-authtool --create-keyring /var/lib/rook/rook-ceph/mon.keyring --gen-key -n mon. --cap mon 'allow *'
2020-01-21 11:23:51.292641 I | exec: Running command: ceph-authtool --create-keyring /var/lib/rook/rook-ceph/client.admin.keyring --gen-key -n client.admin --cap mon 'allow *' --cap osd 'allow *' --cap mgr 'allow *' --cap mds 'allow'
2020-01-21 11:23:51.354473 I | op-mon: creating mon secrets for a new cluster
2020-01-21 11:23:51.376400 I | op-mon: saved mon endpoints to config map map[data: maxMonId:-1 mapping:{"node":{}} csi-cluster-config-json:[{"clusterID":"rook-ceph","monitors":[]}]]
2020-01-21 11:23:51.928502 I | cephconfig: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-01-21 11:23:51.928648 I | cephconfig: generated admin config in /var/lib/rook/rook-ceph
2020-01-21 11:23:53.130316 I | op-mon: targeting the mon count 1
2020-01-21 11:23:53.136582 I | op-mon: sched-mon: created canary deployment rook-ceph-mon-a-canary
2020-01-21 11:23:54.328557 I | op-mon: sched-mon: canary monitor deployment rook-ceph-mon-a-canary scheduled to node-3
2020-01-21 11:23:54.328590 I | op-mon: assignmon: mon a assigned to node node-3
2020-01-21 11:23:54.328602 I | op-mon: assignmon: cleaning up canary deployment rook-ceph-mon-a-canary and canary pvc
2020-01-21 11:23:54.328612 I | op-k8sutil: removing deployment rook-ceph-mon-a-canary if it exists
2020-01-21 11:23:54.335361 I | op-k8sutil: Removed deployment rook-ceph-mon-a-canary
2020-01-21 11:23:54.339585 I | op-k8sutil: rook-ceph-mon-a-canary still found. waiting...
2020-01-21 11:23:56.344659 I | op-k8sutil: rook-ceph-mon-a-canary still found. waiting...
2020-01-21 11:23:58.356248 I | op-k8sutil: rook-ceph-mon-a-canary still found. waiting...
2020-01-21 11:24:00.360749 I | op-k8sutil: rook-ceph-mon-a-canary still found. waiting...
2020-01-21 11:24:02.371784 I | op-k8sutil: rook-ceph-mon-a-canary still found. waiting...
2020-01-21 11:24:04.376864 I | op-k8sutil: rook-ceph-mon-a-canary still found. waiting...
2020-01-21 11:24:06.383763 I | op-k8sutil: rook-ceph-mon-a-canary still found. waiting...
2020-01-21 11:24:08.388665 I | op-k8sutil: confirmed rook-ceph-mon-a-canary does not exist
2020-01-21 11:24:08.388700 I | op-mon: creating mon a
2020-01-21 11:24:08.413074 I | op-mon: mon a endpoint are [v2:10.43.57.20:3300,v1:10.43.57.20:6789]
2020-01-21 11:24:08.433856 I | op-mon: saved mon endpoints to config map map[data:a=10.43.57.20:6789 maxMonId:0 mapping:{"node":{"a":{"Name":"node-3","Hostname":"node-3","Address":"172.31.25.117"}}} csi-cluster-config-json:[{"clusterID":"rook-ceph","monitors":["10.43.57.20:6789"]}]]
2020-01-21 11:24:08.450622 I | cephconfig: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-01-21 11:24:08.450804 I | cephconfig: generated admin config in /var/lib/rook/rook-ceph
2020-01-21 11:24:08.460150 I | cephconfig: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-01-21 11:24:08.460308 I | cephconfig: generated admin config in /var/lib/rook/rook-ceph
2020-01-21 11:24:08.464349 I | op-mon: 0 of 1 expected mon deployments exist. creating new deployment(s).
2020-01-21 11:24:08.483304 I | op-mon: waiting for mon quorum with [a]
2020-01-21 11:24:08.491452 I | op-mon: mon a is not yet running
2020-01-21 11:24:08.491480 I | op-mon: mons running: []
2020-01-21 11:24:08.491684 I | exec: Running command: ceph mon_status --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json --out-file /tmp/966324286
```

**Environment**:
* OS (e.g. from /etc/os-release): Ubuntu 18.04 LTS
* Kernel (e.g. `uname -a`): Linux node-4 5.0.0-1028-gcp
* Cloud provider or hardware configuration: 8vCPU, 16GB, 60 GB SSD per node
* Rook version (use `rook version` inside of a Rook Pod): rook/ceph:v1.1.7
* Storage backend version (e.g. for ceph do `ceph -v`): ceph/ceph:v14.2.4-20190917
* Kubernetes version (use `kubectl version`): 1.17
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): custom
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)): N/A

**Comment by travisn** [comment]
The key is that the operator needs to be able to connect to the mons in order to verify that they are in quorum. From the logs it appears the operator cannot complete this verification so there must be some other aspect of the networking that is missing.

**Comment by penumbra23** [comment]
I've managed to bash into the operator and ping all the monitors with their private cluster IPs, so it's reachable at least for ICMP. Are there any other ports or env vars that I should check?

TBH, I think it's more K8s related, because our other app sometimes also has connectivity issues.

**Comment by BlackCetha** [comment]
I also have this problem. It seems like the Operator is unable to create MONs. It starts the canary deployment pods, but they get stuck. The termination of these deployment pods is stuck and the operator gives up on waiting for them. Sometimes a single MON is started, but it is still unresponsive.
After this, a crashcollector pod pops up but fails to start due to a missing secret.

Things I've tried:
- Clean kubeadm install with Calico and k8s 1.17
- Clean kubeadm install with Calico and k8s 1.16
- Clean kubespray install with Calico and k8s 1.17

This doesn't work with the example YAML files from this repo, neither with ones I've written myself. I have also followed the troubleshooting guide on the website and made sure my cluster nodes met the requirements. Pods on different nodes have connectivity.
